test.cpp

```diff
@@ -349,43 +380,23 @@ std::map<std::string, std::function<void()>> construct_tests() {
  };
 
  tests["autograd/optim/sgd"] = []() {
-   // We better be able to learn XOR
    auto model = ContainerList()
      .append(Linear(2, 8).make())
      .append(Linear(8, 1).make())
      .make();
 
    auto optim = SGD(model, 1e-1).momentum(0.9).nesterov().weight_decay(1e-6).make();
+   test_optimizer_xor(optim, model);
+ };
 
-   float running_loss = 1;
-   int epoch = 0;
-   while (running_loss > 0.1) {
-     auto bs = 4U;
-     auto inp = at::CPU(at::kFloat).tensor({bs, 2});
-     auto lab = at::CPU(at::kFloat).tensor({bs});
-     for (auto i = 0U; i < bs; i++) {
-       auto a = std::rand() % 2;
-       auto b = std::rand() % 2;
-       auto c = a ^ b;
-       inp[i][0] = a;
-       inp[i][1] = b;
-       lab[i] = c;
-     }
-     //
-     // forward
-     auto x = Var(inp);
-     auto y = Var(lab, false);
-     for (auto layer : *model) x = layer->forward({x})[0].sigmoid_();
-     Variable loss = at::binary_cross_entropy(x, y);
-
-     optim->zero_grad();
-     backward(loss);
-     optim->step();
+ tests["autograd/optim/adagrad"] = []() {
+   auto model = ContainerList()
+     .append(Linear(2, 8).make())
+     .append(Linear(8, 1).make())
+     .make();
 
-     running_loss = running_loss * 0.99 + loss.data().sum().toCFloat() * 0.01;
-     EXPECT(epoch < 3000);
-     epoch++;
-   }
+   auto optim = Adagrad(model, 1.0).make();
```

Comment:
I would add lr_decay and weight_decay in here as well.

Reply:
Fixed